concepts/http-server/about.md

```diff
@@ -0,0 +1,108 @@
+~~~~
+
+### Handlers
+
+```go
+type Handler interface {
+ ServeHTTP(ResponseWriter, *Request)
+}
+```
+
+Handlers are responsible for satisfying the client's requests by performing various tasks. Each object that implements the `http.Handler` interface, is capable of responding to client requests. Handler functions receive an `http.Request` which contains information regarding the client’s request, and an `http.ResponseWriter` that allows the handler to write the appropriate response in accordance with the request. `http.ResponseWriter` interface contains a `Header` field (!!! I need to complete this !!!), a `Write([]byte) (int, error)` method that writes the response to the client, and a `WriteHeader(statusCode int)`  that writes the appropriate status code to the client.
+
+~~~~exercism/caution
+If you wish to send and status code besides `http.StatusOk` you need to use `WriteHeader` before `Write`, since `Write` assumes `http.StatusOk` if no status is set before invoking it.  
+Example:  
+```go
+func(w http.ResponseWriter, r *http.Request){
+ w.Write([]byte("Not Found!"))
+ w.WriteHeader(http.StatusNotFound)
+}
+// The result of running this function is a response with status code 200 Ok!
+----------
+func(w http.ResponseWriter, r *http.Request){
+ w.WriteHeader(http.StatusNotFound)
+ w.Write([]byte("Not Found!"))
+}
+// The result of running this function is a response with status code 404 NotFound!
+```
+~~~~
+
+### Multiplexer
+
+`http.ServeMux` is responsible for routing each request to the right handler by matching the pattern in the request with the **longest** pattern specified for it. `http.NewServeMux() *ServeMux` is used to create a new multiplexer. It uses handlers to route each request to the proper handler. As an example:
+```go
+func multiplexer() *ServeMux{
+ serveMux := http.NewServeMux()
+ serveMux := http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request){
+ w.Write([]byte("Hello World!"))
+ })
+ return serveMux
+}
+```
+
+#### Absolute paths vs subtrees
+
+Paths that end without a `/` are called absolute paths (e.g. `"root/image"`). If an absolute path is used in the request, the multiplexer will try to find the “exact” match for that path. On the contrary, paths that end with `/` are called subtrees (e.g. `“root/image/“`) and the multiplexer will try to find the best match for these paths. Beware that the multiplexer will convert an absolute path to a subtree if it does not find a corresponding handler.  
+
+~~~~exercism/caution
+Draining the request body before closing it:  
+Unlike the client that drains the body before closing it, in the server you need to manually drain the request body. A good way to do so is writing a middleware that closes the request after the server job is done (I have seen this pattern for the first time in [Adam Woodbeck's mux_test.go][awoodbeck-gnp-chapter09-mux_test.go]):   
```

Comment:
I have never heard of this and I don't think it is correct.

The official documentation https://pkg.go.dev/net/http#Request states
```
	// For server requests, the Request Body is always non-nil
	// but will return EOF immediately when no body is present.
	// The Server will close the request body. The ServeHTTP
	// Handler does not need to.
```

The behavior here is different then on the client side where you need to close the response body on your own.

Reply:
@junedev 
Ok so after a lot of digging I did not find an answer that satisfies me regarding this issue.
First please take a look at this:
http://tleyden.github.io/blog/2016/11/21/tuning-the-go-http-client-library-for-load-testing/
I have tested it with go version 1.19.3 and it still behaves the same.
The response.Body.Close() will not drain the body therefore not allowing the same connection to get reused.
This is also another post discussing the same issue in 2017:
https://forum.golangbridge.org/t/do-i-need-to-read-the-body-before-close-it/5594

> // The response body is streamed on demand as the Body field
> 	// is read. If the network connection fails or the server
> 	// terminates the response, Body.Read calls return an error.
> 	//
> 	// The http Client and Transport guarantee that Body is always
> 	// non-nil, even on responses without a body or responses with
> 	// a zero-length body. It is the caller's responsibility to
> 	// close Body. The default HTTP client's Transport may not
> 	// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
> 	// not read to completion and closed.

The same principal I think applies for the r *http.Request since both bodies are the same type.
I don't know if I am missing something here but if this is true, then we should drain the client body as well.